# Homework 03

## Due Thursday, February 18th

Yoonso Kang

In [50]:
import numpy as np

### 1. Vector manipulation

Create a vector of zeros of size 10 but have the fifth value be 1.

In [2]:
z = np.zeros(10)
z[4] = 1
print(z)

[ 0.  0.  0.  0.  1.  0.  0.  0.  0.  0.]


### 2. Vector range

Create a vector with values ranging from 10 to 49, inclusive.

In [3]:
np.linspace(10, 49, 40)

array([ 10.,  11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,
        21.,  22.,  23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,
        32.,  33.,  34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,
        43.,  44.,  45.,  46.,  47.,  48.,  49.])

### 3. Reverse a vector

Write a function, `reverse(x)` which reverses a vector, so that the first element becomes the last element. The API should be

    return a 1D vector with the elements reversed
    
    Input
    -----
        x: 1D array
    
    Output
    ------
        res: 1D array, with the elements of `x` reversed

Include this (or an equivalent description) in the function's docstring.

In [19]:
def reverse(x):
    """Returns a 1D array with the elements reversed.
    
    Input:
        x: 1D array
    
    Output:
        res: 1D array, with the elements of `x` reversed
    """
    return x[::-1]

reverse?
print(reverse(z)) # From question 1
print(reverse(np.linspace(10, 49, 40))) # From question 2

assert (reverse(reverse(z)) == z).all()

[ 0.  0.  0.  0.  0.  1.  0.  0.  0.  0.]
[ 49.  48.  47.  46.  45.  44.  43.  42.  41.  40.  39.  38.  37.  36.  35.
  34.  33.  32.  31.  30.  29.  28.  27.  26.  25.  24.  23.  22.  21.  20.
  19.  18.  17.  16.  15.  14.  13.  12.  11.  10.]


### 4. Find indices of non-zero elements

Write a function `find_nonzero(x)` that finds the indices of non-zero elements, with an API like

    return the indices of non-zero elements of an array.

    Input:      x:  array
    Output:     indices: array of integer indices

Use the function to get out the non-zero elements from an array containing the elements [1,2,0,0,4,0].

In [20]:
def find_nonzero(x):
    """Returns the indices of non-zero elements of an array.
    
    Input:
        x: array
    
    Output:
        indices: array of integer indices
    """
    return np.where(x != 0)

x = np.array([1, 2, 0, 0, 4, 0])
print(x[find_nonzero(x)])

y = np.tile(x, (5, 1)) # Another array containing the elements [1, 2, 0, 0, 4, 0]
print(y[find_nonzero(y)])

[1 2 4]
[1 2 4 1 2 4 1 2 4 1 2 4 1 2 4]


### 5. Array with random values

Create a 3x3x3 array with normally distributed ($\mu =0, \sigma=1)$ random values.

In [9]:
np.random.randn(3, 3, 3)

array([[[-1.23748918,  1.25465982, -0.80228455],
        [-0.38417395,  0.20477038, -0.87735327],
        [ 1.02679909, -0.4208172 ,  0.12930184]],

       [[ 0.18331097,  0.79213949, -1.31452986],
        [-0.08691685,  1.82103661, -1.64417132],
        [-0.02706155,  0.87736085, -0.37520413]],

       [[-0.46783498,  0.34556798,  0.4718769 ],
        [ 0.26483447,  0.15503641,  0.0595081 ],
        [-0.08978065,  0.70168569, -0.01170467]]])

### 6. Operations on an array with random values

Create a 10x10 array with random values. Find 
- the minimum value of the 2nd row
- the maximum value of the 4th column

In [23]:
c = np.random.rand(10, 10)

print('The minimum value of the 2nd row =', c[1, :].min())
print('The maximum values of the 4th row =', c[:, 3].max())
c

The minimum value of the 2nd row = 0.0427729978757
The maximum values of the 4th row = 0.952130518393


array([[ 0.31150091,  0.72015492,  0.18022303,  0.95213052,  0.06425199,
         0.34467702,  0.88042681,  0.52286171,  0.70403648,  0.95924046],
       [ 0.31517409,  0.88475528,  0.45260164,  0.78403337,  0.6472188 ,
         0.21470667,  0.19666586,  0.62983071,  0.218877  ,  0.042773  ],
       [ 0.96415547,  0.42309613,  0.15992285,  0.17390219,  0.71982004,
         0.36142739,  0.02170152,  0.55204013,  0.89092729,  0.53182499],
       [ 0.12312617,  0.62613971,  0.53261328,  0.18422765,  0.95612804,
         0.02135663,  0.62891084,  0.40087738,  0.42538064,  0.02756474],
       [ 0.81814448,  0.09257613,  0.78746038,  0.02140809,  0.40608641,
         0.41703992,  0.35309777,  0.30073252,  0.0078728 ,  0.11334038],
       [ 0.73857514,  0.07259271,  0.73687782,  0.62133183,  0.34002754,
         0.15317115,  0.04443163,  0.59478516,  0.8347467 ,  0.83444039],
       [ 0.44685705,  0.82114691,  0.50667135,  0.04961717,  0.70979853,
         0.06619581,  0.41619349,  0.73395876

### 7. Vector without end points

Create a vector of size 10 with values ranging from 0 to 1, both excluded.

In [11]:
np.linspace(1, 10, 10)/11

array([ 0.09090909,  0.18181818,  0.27272727,  0.36363636,  0.45454545,
        0.54545455,  0.63636364,  0.72727273,  0.81818182,  0.90909091])

### 8. Sorted random vector

Create a random vector of size 10, then sort it.

In [18]:
d = np.random.rand(10)

print('The original vector =', d)
print('The sorted vector =', np.sort(d))

The original vector = [ 0.61796491  0.90819333  0.67897666  0.40011318  0.40853488  0.3917514
  0.71678647  0.35763299  0.74750749  0.20416891]
The sorted vector = [ 0.20416891  0.35763299  0.3917514   0.40011318  0.40853488  0.61796491
  0.67897666  0.71678647  0.74750749  0.90819333]


### 9. Test for array equality

Write a function `is_equal(A, B)` that compare two arrays `A` and `B` and checks to see if they are equal. 

    return True if two arrays are the same shape and all the elements are equal.
    
    Inputs:
        A, B:  array
    Output:
        Bool:  True if A and B have the same shape and identical elements.

The arrays count as equal if they are the same shape and all elements in the array are equal.

In [21]:
def is_equal(A, B):
    """Returns True if two arrays are the same shape and all the elements are equal.
    
    Inputs:
        A, B:  array
    
    Output:
        Bool:  True if A and B have the same shape and identical elements.
    """
    return (A.shape == B.shape) & (A.flat == B.flat).all()

print(is_equal(z, reverse(z)))
print(is_equal(y, y.flatten()))

False
False


Or, one of the functions that are provided by the `numpy` package can be used here instead.

In [22]:
def is_equal(A, B):
    """Returns True if two arrays are the same shape and all the elements are equal.
    
    Inputs:
        A, B:  array
    
    Output:
        Bool:  True if A and B have the same shape and identical elements.
    """
    return np.array_equal(A, B)

print(is_equal(z, reverse(z)))
print(is_equal(y, y.flatten()))

False
False


### 10. Subtract the row mean from rows of a matrix

Create a matrix. Then subtract the mean of each row of the matrix from the respective row. Do this without using a loop (that is, using array operations).

In [24]:
e = np.random.randint(10, size=(5, 10))

print(e-e.mean(axis=1)[:, np.newaxis])
print('The original matrix =\n', e)

[[ 3.1 -3.9 -3.9  0.1  0.1  4.1  2.1  0.1  1.1 -2.9]
 [ 2.3  3.3  2.3  3.3 -3.7 -1.7 -0.7 -0.7 -2.7 -1.7]
 [ 0.7  2.7  2.7  2.7 -5.3 -5.3 -2.3  2.7  0.7  0.7]
 [-1.1  5.9 -3.1 -2.1 -3.1  0.9  0.9  0.9 -2.1  2.9]
 [-1.3  0.7  5.7 -2.3  0.7 -2.3 -2.3  4.7 -3.3 -0.3]]
The original matrix =
 [[8 1 1 5 5 9 7 5 6 2]
 [6 7 6 7 0 2 3 3 1 2]
 [7 9 9 9 1 1 4 9 7 7]
 [2 9 0 1 0 4 4 4 1 6]
 [2 4 9 1 4 1 1 8 0 3]]


### 11. Multiply arrays

Multiply an array of dimension (5,5,3) with an array with dimensions (5,5). Then multiply an array of dimension (3,5,5) with an array of dimension (5,5).

In [26]:
f = np.arange(5*5*3).reshape(5, 5, 3)
g = np.arange(5*5).reshape(5, 5)
h = np.arange(3*5*5).reshape(3, 5, 5)

print(f.shape, 'x', g.shape, '->', (f * g[:, :, np.newaxis]).shape, '\n', f * g[:, :, np.newaxis])
print(h.shape, 'x', g.shape, '->', (h * g).shape, '\n', h * g)

(5, 5, 3) x (5, 5) -> (5, 5, 3) 
 [[[   0    0    0]
  [   3    4    5]
  [  12   14   16]
  [  27   30   33]
  [  48   52   56]]

 [[  75   80   85]
  [ 108  114  120]
  [ 147  154  161]
  [ 192  200  208]
  [ 243  252  261]]

 [[ 300  310  320]
  [ 363  374  385]
  [ 432  444  456]
  [ 507  520  533]
  [ 588  602  616]]

 [[ 675  690  705]
  [ 768  784  800]
  [ 867  884  901]
  [ 972  990 1008]
  [1083 1102 1121]]

 [[1200 1220 1240]
  [1323 1344 1365]
  [1452 1474 1496]
  [1587 1610 1633]
  [1728 1752 1776]]]
(3, 5, 5) x (5, 5) -> (3, 5, 5) 
 [[[   0    1    4    9   16]
  [  25   36   49   64   81]
  [ 100  121  144  169  196]
  [ 225  256  289  324  361]
  [ 400  441  484  529  576]]

 [[   0   26   54   84  116]
  [ 150  186  224  264  306]
  [ 350  396  444  494  546]
  [ 600  656  714  774  836]
  [ 900  966 1034 1104 1176]]

 [[   0   51  104  159  216]
  [ 275  336  399  464  531]
  [ 600  671  744  819  896]
  [ 975 1056 1139 1224 1311]
  [1400 1491 1584 1679 1776]]]


### 12. Creating a new array

Starting with an array

    Z = [1,2,3,4,5,6,7,8,9,10,11,12,13,14].
    
Generate an array 

    R = [[1,2,3,4], [2,3,4,5], [3,4,5,6], ..., [11,12,13,14]]. 

In [27]:
Z = np.arange(1, 15)

* The first method using the function `np.ones`

In [29]:
R = Z[:11, np.newaxis] + Z[np.newaxis, :4] - np.ones((11, 4), dtype=int)
print('R =\n', R)

R =
 [[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]


* The second and third methods without using the function `np.ones`

In [30]:
R = (Z-1)[:11, np.newaxis] + Z[np.newaxis, :4]
print('R =\n', R)

R =
 [[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]


In [31]:
R = Z[:11, np.newaxis] + (Z-1)[np.newaxis, :4]
print('R =\n', R)

R =
 [[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]


### 13. Convert from Cartesian to polar coordinates

Write a function `cart2polar(xy)` which takes in an Nx2 matrix of `x, y` pairs representing Cartesian coordinates and return the locations represented in polar coordinates in the same shape matrix. The API is

    Return equivalent polar coordinates of specified cartesian points.
    
    Inputs:
        xy: array
             Nx2 array of Cartesian coordinates, with columns x and y respectively.

    Returns:
        rp: array
            Nx2 array of polar coordinates, with columns radius and angle respectively


In [33]:
def cart2polar(xy):
    """Returns equivalent polar coordinates of specified cartesian points.
    
    Inputs:
        xy: array
            Nx2 array of Cartesian coordinates, with columns x and y respectively.
    
    Returns:
        rp: array
            Nx2 array of polar coordinates, with columns radius and angle respectively.
    """
    r = np.sqrt(np.square(xy).sum(axis=1))
    p = np.angle(np.dot(xy, np.array([1., 1.j])), deg=True)
    p[p < 0] = p[p < 0] + 360.
    rp = np.array([[rn, pn] for rn, pn in zip(r, p)]) # Or, `rp = np.vstack((r, p)).T`
    return rp

o = np.random.randint(-10, 10, size=(15, 2)) + np.random.rand(15, 2)
print('[x, y] =\n', o)
print('[r, phi(degree)] =\n', cart2polar(o))

[x, y] =
 [[ 4.70271677  9.52423484]
 [-5.27555847 -8.19325891]
 [-4.80589352  9.78637631]
 [-7.41525878 -6.9116391 ]
 [ 6.32599684  5.54168906]
 [-8.13813509  5.90377594]
 [-3.89449209  8.43139637]
 [-6.8579456  -9.23235896]
 [-3.79850967  8.57963907]
 [ 9.93856809  3.3953836 ]
 [ 9.99801689  5.55249573]
 [ 3.86722616  8.30466953]
 [ 5.79810154  7.12262372]
 [ 1.71010968 -8.22897681]
 [-6.36745926  1.97780985]]
[r, phi(degree)] =
 [[  10.62198637   63.7215348 ]
 [   9.74479393  237.222915  ]
 [  10.90274158  116.15473072]
 [  10.13690376  222.98676546]
 [   8.41002697   41.2189534 ]
 [  10.05404462  144.04114931]
 [   9.28738464  114.79238381]
 [  11.50077692  233.39443766]
 [   9.38290373  113.88066653]
 [  10.50255995   18.86207067]
 [  11.43636965   29.04602861]
 [   9.16094832   65.0300698 ]
 [   9.18421201   50.85302776]
 [   8.40479235  281.73985548]
 [   6.66755346  162.74457376]]


### 14. Compute and time for loops and array operations.

Calculate the time it takes to do what we did in class

    a = np.arange(1024.0).reshape(4, 8, 16, 2)   # a 4D array using sequential numbers
    b = np.random.rand(4, 8, 16, 2)              # a 4D array using random numbers
    sol = a * b  # element-by-element multiplication. This operation is about as fast as it can be on your computer.

and the time required for doing the same operation with a series of 4 `for` loops, one for each dimension of the arrays. Compare the times by calculating a ratio.

In [35]:
ebye = %timeit -o a = np.arange(1024.0).reshape(4, 8, 16, 2); b = np.random.rand(4, 8, 16, 2);\
 sol = a * b

10000 loops, best of 3: 67.2 µs per loop


In [36]:
%%timeit -n10000 -p18
a = np.arange(1024.0).reshape(4, 8, 16, 2)
b = np.random.rand(4, 8, 16, 2)
sol = np.empty(a.shape)
for i in np.arange(a.shape[0]):
    for j in np.arange(a.shape[1]):
        for k in np.arange(a.shape[2]):
            for l in np.arange(a.shape[3]):
                sol[i, j, k, l] = a[i, j, k, l] * b[i, j, k, l]


10000 loops, best of 3: 9.24985670780297298 ms per loop


In [37]:
ratio = 9.24985670780297298e-03/ebye.best
ratio

137.56277517416225

Therefore, the element-by-element multiplication is approxiately 138 times faster than the multiplication using `for` loops.

* I tried to also capture the output of a series of 4 `for` loops like the element-by-element multiplication in the first one, but the options somehow do not work as the same way for the timeit cell magic. (They have the same options.) To save the output into a variable, I created a fuction containing the same procedures.

In [38]:
def for4(a, b):
    sol = np.empty(a.shape)
    for i in np.arange(a.shape[0]):
        for j in np.arange(a.shape[1]):
            for k in np.arange(a.shape[2]):
                for l in np.arange(a.shape[3]):
                    sol[i, j, k, l] = a[i, j, k, l] * b[i, j, k, l]
    return sol

frlp = %timeit -n10000 -o a = np.arange(1024.0).reshape(4, 8, 16, 2); \
b = np.random.rand(4, 8, 16, 2); for4(a, b)

10000 loops, best of 3: 8.85 ms per loop


In [39]:
ratio = frlp.best/ebye.best
ratio

131.5563666003892

With the above fuction, it took about 132 times longer for the series of `for` loops to do the same thing as the element-by-element multiplication.
* This time, it has turns out that this took shorter than the case without the function. But sometimes, vice versa.

### 15. Fix the code

This code needs to be fixed so that it can run properly and pass the assertion test at the bottom. Write out what you changed to make it work and why it needed to be changed.

In [ ]:
def calculate(a, b):
    """Do some calculations
    
    Take in two arrays, subtract 1/4 of the [:2,:] elements from the elements themselves.
    Then multiply with array b and return the result.
    
    Inputs:
        a: input array, shape [MxN]
        b: input array, shape [M]
    
    Returns:
        resultant array
    """

    aa = a[:2, :]
    aa /= 4.
    a[:2, :] -= aa

    return a*b
    

a = np.ones((3, 2))
b = np.ones(3)*3
res = calculate(a, b)

assert (res == np.array([[ 2.25,  2.25],[ 2.25,  2.25],[ 3., 3.]])).all()

#### A list of changes

1. Line 19: `a*b` &rarr; `a*b[:, np.newaxis]` Since the shape of array `a` is `(M, N)` and the shape of array `b` is `(M, )`, the shape of `b` array needs to become `(M, 1)` for the multiplication.
2. Line 15: `aa = a[:2, :]` &rarr; `aa = a.copy()[:2, :]` Since subsets of an array have the same memory as the original array, a `copy` of the array is needed to have its own memory if the subset is going to be modified like in the next line.
3. Lines 14, 18, and 20: Unnecessary lines that have nothing to do with making the above code work are deleted.

In [40]:
def calculate(a, b):
    """Do some calculations
    
    Take in two arrays, subtract 1/4 of the [:2,:] elements from the elements themselves.
    Then multiply with array b and return the result.
    
    Inputs:
        a: input array, shape [MxN]
        b: input array, shape [M]
    
    Returns:
        resultant array
    """
    aa = a.copy()[:2, :]
    aa /= 4.
    a[:2, :] -= aa
    return a*b[:, np.newaxis]

a = np.ones((3, 2))
b = np.ones(3)*3
res = calculate(a, b)

assert (res == np.array([[ 2.25,  2.25],[ 2.25,  2.25],[ 3., 3.]])).all()

### 16. Calculate components of a vector

Read in file `data/burl1h2010.txt`. This file gives a bunch of data from a wind station. In particular, read in the wind direction, `WDIR`, and wind speed, `WSPD`. Then, write a function that calculates and returns the `u` and `v` components of this wind vector, which point along the x- and y-axes.

The function should have the following basic functionality:

    Take in vector and return components.
    
    Inputs:
        vdir: vector direction in degrees
        vspd: vector magnitude
        
    Returns:
        u: x component of vector
        v: y component of vector


In [51]:
wdir, wspd = np.loadtxt('../python4geosciences/data/burl1h2010.txt', usecols=[5, 6], unpack=True)

In [52]:
def windvector(vdir, vspd):
    """Takes in vector and return components.
    
    Inputs:
        vdir: vector direction in degrees
        vspd: vector magnitude
    
    Returns:
        u: x component of vector
        v: y component of vector
    """
    u = vspd*np.cos(vdir*np.pi/180.)
    v = vspd*np.sin(vdir*np.pi/180.)
    return u, v

u, v = windvector(wdir, wspd)
print('[u, v] =\n', np.array([[un, vn] for un, vn in zip(u, v)]))

np.allclose(cart2polar(np.vstack((u, v)).T), np.vstack((wspd, wdir)).T)

[u, v] =
 [[ -2.08397511   2.1580194 ]
 [ -1.96070053   3.01921404]
 [ -2.12119911   2.52794666]
 ..., 
 [ -9.56084447   6.208885  ]
 [-10.50707636   5.5867116 ]
 [ -8.83345912   5.1       ]]


True

### 17. (Grad) Write and use a class

Consider one of the data files located in `python4geosciences/data/MCH/`. Write a class, `dataclass`, to store some of the information in the file. The class should:

* be initialized with the name of the file (without file type suffix of .cnv2), and the temperature and salinity values found in the file;
* contain a method, `salt_mean`, which returns the mean of the salinity;
* contain a method, `temp_max`, which returns the maximum value of the temperature values.

Notice that there are many data files in this directory. It would be a lot of work to read them all in by hand. Instead, write a loop which loops through the file names, reads in the necessary data, calls `dataclass` to store the data in a class instance, and adds it to a list of all of the class instances. At the end of your solution, you should have a list of all of the class instances, one for each data file. As usage examples, if your list of data objects is called `wind`, you should be able to access the temperatures from the 3rd file that is read in with

    wind[2].temp

and you should be able to calculate the mean salinity of the final file with 

    wind[-1].salt_mean()

Hint: Check out the package `glob` for reading in file names with a wild card. 

In [44]:
class dataclass(object):
    
    def __init__(self, filename, temp, salt):
        self.filename = filename
        self.temp = temp
        self.salt = salt
    
    def salt_mean(self):
        'Returns the mean of the salinity.'
        return sum(self.salt)/len(self.salt)
    
    def temp_max(self):
        'Returns the maximum value of the temperature values.'
        return max(self.temp)


In [46]:
import glob
import codecs

MCH = []

for pathname in glob.glob('../python4geosciences/data/MCH/*.cnv2'):
    f = codecs.open(pathname, 'r', 'UTF-8')
    f.seek(0)
    
    filename = '.'.join(pathname.split('/')[-1].split('.')[:-1])
    temp = []
    salt = []
   
    for line in f.readlines():
        if list(line)[0] == ' ':
            data = line.split()
            temp.append(float(data[2]))
            salt.append(float(data[5]))
        else:
            continue
    MCH.append(dataclass(filename, temp, salt))

print('The temperature from the 3rd file ', MCH[2].filename, '.cnv2 =\n', MCH[2].temp, sep='')
print('The mean salinity of the final file ', MCH[-1].filename, '.cnv2 = ', \
      MCH[-1].salt_mean(), sep='')

The temperature from the 3rd file m1002a.cnv2 =
[21.2128, 21.2149, 21.2177, 21.219, 21.2165, 21.2331, 21.3468, 21.4867, 21.526, 21.5914, 21.6244, 21.6219, 21.6295, 21.6204, 21.6003, 21.5912, 21.4888, 21.3504, 21.2779, 21.343, 21.4117, 21.4415, 21.441, 21.3959, 21.3208, 21.2352, 21.1793, 21.1591, 21.1479, 21.1501, 21.1435, 21.1413, 21.1385, 21.1377, 21.1369]
The mean salinity of the final file m1018b.cnv2 = 35.828875000000004


### 18. (Grad) Running mean

Write a function that computes averages using a sliding window over a vector; that is, it should take in a vector and for every element in the vector, compute the average of that point with the N points on either side of the element. N should also be an input parameter, and the function should return the original vector if N=0. The returned vector should be the same size as the input vector.

Do not use the `pandas` package or the function `convolve`; instead, use a loop. Use the following API for your function:

    Return the running average of a vector, with averaging window of size N.
    
    Inputs:
        vec: Input vector
        N (int): Averaging window size, or number of elements to include in mean on either side of each element.
    
    Outputs:
        out: Vector of running-averaged values; same size as vec


In [48]:
def running_mean(vec, N):
    """Returns the running average of a vector, with averaging window of size N.
    
    Inputs:
        vec: Input vector
        N (int): Averaging window size, or number of elements to include in mean on either side of each element.
    
    Outputs:
        out: Vector of running-averaged values; same size as vec
    """
    out = -999*np.ones(vec.size)
    if N == 0:
        out = vec
    else:
        for m in np.arange(vec.size-N+1):
            out[np.arange(m, m+N).mean(dtype=int)] = vec[m:m+N].mean()
    return np.ma.masked_where(out == -999, out)

w = np.arange(15.)

for n in np.arange(18):
    print('N =', n, running_mean(w, n))

N = 0 [  0.   1.   2.   3.   4.   5.   6.   7.   8.   9.  10.  11.  12.  13.  14.]
N = 1 [  0.   1.   2.   3.   4.   5.   6.   7.   8.   9.  10.  11.  12.  13.  14.]
N = 2 [0.5 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5 10.5 11.5 12.5 13.5 --]
N = 3 [-- 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 13.0 --]
N = 4 [-- 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5 10.5 11.5 12.5 -- --]
N = 5 [-- -- 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 -- --]
N = 6 [-- -- 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5 10.5 11.5 -- -- --]
N = 7 [-- -- -- 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 -- -- --]
N = 8 [-- -- -- 3.5 4.5 5.5 6.5 7.5 8.5 9.5 10.5 -- -- -- --]
N = 9 [-- -- -- -- 4.0 5.0 6.0 7.0 8.0 9.0 10.0 -- -- -- --]
N = 10 [-- -- -- -- 4.5 5.5 6.5 7.5 8.5 9.5 -- -- -- -- --]
N = 11 [-- -- -- -- -- 5.0 6.0 7.0 8.0 9.0 -- -- -- -- --]
N = 12 [-- -- -- -- -- 5.5 6.5 7.5 8.5 -- -- -- -- -- --]
N = 13 [-- -- -- -- -- -- 6.0 7.0 8.0 -- -- -- -- -- --]
N = 14 [-- -- -- -- -- -- 6.5 7.5 -- -- -- -- -- -- --]
N = 15 [-- -- --